source/_posts/2018-01-12-pager-duty-and-confluence.markdown

```diff
@@ -0,0 +1,251 @@
+---
+layout: post
+title: "PagerDuty and Confluence"
+date: 2018-01-12 18:19:49 +0000
+comments: true
+author: "Bryan Kroger bryan.kroger@grandrounds.com"
+categories: automation
+---
+My first project here at Grand Rounds was to pick up one of my favorite topics of IT and operations 
+in general: on-call rotations!  Making sure we have accurate alarming and alerting never seems like the
+most glamorous work in the world, but it is often very important and highly impactful.
+
+To that end, as with most things, we want to start with a data-driven approach to identifying and resolving
+our problems.  This is a pattern that I've used several times over the years for a fun little hack which
+allows us to build reports based on PagerDuty incident data.
+
+This gives us two graphs:
+
+* Number of incidents per service.
+* Amount of time each service spends in incident ( on average ).
+
+We fully expect that this code is a little rough around the edges, and that's okay.  We're building things
+here that are going to give us ideas and insight into our bigger picture items.  Ideally we want to
+spend as little time as possible gold plating this code because we're not 100% sure this will be something
+that we intend to live forever.  Just like in the startup business world we want to create the Minimal
+Viable Product ( MVP ) which allows us to prove out a business case.
+
+In this case we will prove out the value of incident reporting data.  This is a perfect example of a product
+that is valuable today, but becomes less valuable over time because it allows us to refine a process.
+
+We start this with a high-level HTTPClient.  Our nickname for this is called Groot: 
+Grand Rounds Operational and Organizational Tool.  We're all fans of the comic books over here, 
+so we often name our projects after Marvel characters.  I haven't seen anything from the DC 
+realm quite yet, but maybe someday.
+
+```ruby
+class GrootHTTPClient
+  @url
+  @headers
```

Comment:
Was your intention here to create `@url` and `@headers` instance variables on the `GrootHTTPClient` object, in addition to creating similarly named instance variables on `GrootHTTPClient` instances below in the `#initialize` method? I'm wondering if you instead intended to "declare" the instance-level instance variables at the top of the class definition in the way one would in Java, for example. In Ruby this is not necessary.

In fact, it is not even possible. Ruby treats classes as objects, i.e. as instances of a class, so classes themselves can have instance variables too:
```ruby
class Foo                                                                                                                                                                     
  @var = "instance variable on #{self} which is an instance of #{self.class}"                                                                                                 
                                                                                                                                                                              
  def initialize                                                                                                                                                              
    @var = "instance variable on #{self} which is an instance of #{self.class}"                                                                                               
  end                                                                                                                                                                         
end                                                                                                                                                                           
                                                                                                                                                                              
Foo.instance_variable_get(:@var) #=> "instance variable on Foo which is an instance of Class"                                                                                 
Foo.new.instance_variable_get(:@var) #=> "instance variable on #<Foo:0x00000015b1eaa8> which is an instance of Foo"
```
In your case, since you haven't assigned any value to the class-level `@url` variable, the result of
```ruby
GrootHTTPClient.instance_variable_get(:@url)
```
will just be nil, which is harmless. Then again, it's not great style to reference `@url` at the top of the class without assigning anything to it or ever using it later, and since this otherwise excellent blog post will enter the wild -- where it could attract or repel potential employees -- I think style matters a little bit.

Reply:
I'm not sure what this means.  I think perhaps your ruby-foo is much better than mine.  Would you be willing to help me fix the code so we can get it to a place that is where you'd want it?

Reply:
Haha, sorry for not being very clear. I was just suggesting to remove lines 38, 39, 59, 60, and 128.

Reply:
Ohh, but...but...I thought we had to have those thingies there?  Isn't that good design?

Reply:
>Ohh, but...but...I thought we had to have those thingies there? Isn't that good design?

As I understand it, listing these attributes when not modifying them is not common practice in Ruby. You can also see this in e.g. ActiveModel subclasses, which may have many attributes, backed by database columns, that are not explicitly listed in the model class definition. It's common practice to have to look in the e.g. `structure.sql` file, or inspect an actual instance of the model, to identify what attributes are present.

Reply:
I don't want this to linger, but we have until Tuesday to push this.

Reply:
This ain't a blocker!

Reply:
I was only pointing out that whereas in a language like Java you would do
```java
public class Foo {
  private int var;

  public Foo(int var) {
    this.var = var;
  }
}
```
in Ruby you would just do
```ruby
class Foo
  def initialize(var)
    @var = var
  end
end
```
but not
```ruby
class Foo
  @var

  def initialize(var)
    @var=var
  end
end
```
since in Ruby one does not declare instance variables in the way that one does in a language like Java.